modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/container/request/filter/NestedFieldsContainerRequestFilter.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.object.rest.internal.jaxrs.container.request.filter;
+
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectRelationship;
+import com.liferay.object.service.ObjectRelationshipLocalService;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.util.PropsValues;
+import com.liferay.portal.vulcan.fields.NestedFieldsContext;
+import com.liferay.portal.vulcan.fields.NestedFieldsContextThreadLocal;
+
+import jakarta.ws.rs.container.ContainerRequestContext;
+import jakarta.ws.rs.container.ContainerRequestFilter;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.ext.Provider;
+
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * @author Carlos Correa
+ */
+@Provider
+public class NestedFieldsContainerRequestFilter
+	implements ContainerRequestFilter {
+
+	public NestedFieldsContainerRequestFilter(
+		ObjectRelationshipLocalService objectRelationshipLocalService) {
+
+		_objectRelationshipLocalService = objectRelationshipLocalService;
+	}
+
+	@Override
+	public void filter(ContainerRequestContext containerRequestContext)
+		throws IOException {
+
+		NestedFieldsContext nestedFieldsContext =
+			NestedFieldsContextThreadLocal.getNestedFieldsContext();
+
+		if (nestedFieldsContext == null) {
+			return;
+		}
+
+		List<String> nestedFields = new ArrayList<>(
+			nestedFieldsContext.getNestedFields());
+
+		if (!nestedFields.contains("rootModelHierarchy")) {
+			return;
+		}
+
+		Queue<Long> queue = new LinkedList<>();
+
+		queue.add(_objectDefinition.getObjectDefinitionId());
+
+		Set<Long> processed = new HashSet<>();
+
+		Long objectDefinitionId = null;
+
+		while ((objectDefinitionId = queue.poll()) != null) {
+			if (!processed.add(objectDefinitionId)) {
+				continue;
+			}
+
+			for (ObjectRelationship objectRelationship :
+					_objectRelationshipLocalService.getObjectRelationships(
+						objectDefinitionId, true)) {
+
+				nestedFields.add(objectRelationship.getName());
+
+				if (objectRelationship.getObjectDefinitionId1() !=
+						objectDefinitionId) {
+
+					queue.add(objectRelationship.getObjectDefinitionId1());
+				}
+				else if (objectRelationship.getObjectDefinitionId2() !=
+							objectDefinitionId) {
+
+					queue.add(objectRelationship.getObjectDefinitionId2());
+				}
+			}
+		}
```

Comment:
Hi @ccorreagg ,

@pedro-oliveira446 updated this logic to use the `ObjectDefinitionTreeFactory` for creating the root model tree, ensuring consistency with [this logic.](https://github.com/carolmariaabb/liferay-portal/blob/2ff83884c136494de528e84602ca2762aa012ca3/modules/apps/object/object-api/src/main/java/com/liferay/object/tree/ObjectDefinitionTreeFactory.java#L50-L65)

Could you please check?

Reply:
you can see the change [here](https://github.com/pedro-oliveira446/liferay-portal/commit/2344d245697eed06082f6d0e7100fefff2467107#diff-36cd0558f894cc2e6e345052868642afebca9f795a1359d1d4c950b007d8d2a3R70-R105)